### PR TITLE
OCPBUGS-38289: fix(prometheus): avoid passing the cluster-wide proxy config to a remote endpoint when noProxy will prevent the proxuying anyway

### PR DESCRIPTION
**Pull request by machine424**
Also

docs: hide PrometheusK8sConfig.RemoteWrite.ProxyURL from docs until its usage is clarified

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: pkg/manifests/types.go [hunk 1]
```diff
@@ -757,7 +757,7 @@ type RemoteWriteSpec struct {
 	Name string `json:"name,omitempty"`
 	// Defines OAuth2 authentication settings for the remote write endpoint.
 	OAuth2 *monv1.OAuth2 `json:"oauth2,omitempty"`
-	// Defines an optional proxy URL.
```
Comment (machine424) [inline pkg/manifests/types.go:760]: This should be reflected in the docs as well.
(Tell me if we should keep this to another PR)

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -1424,6 +1425,29 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, telemetrySecret *v1.Secret)
...
+		proxyFunc := proxyConfig.ProxyFunc()
+		// Ignore errors here.
+		u, _ := url.Parse(rw.URL)
+		proxyURL, err := proxyFunc(u)
+		if err == nil && proxyURL == nil {
+			// Proxing will not be used, no need to set it.
+			continue
+		}
+
 		if f.proxy.HTTPProxy() != "" {
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1451]: hmm shouldn't we remove this code and use `proxyURL` instead?

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -1424,6 +1425,29 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, telemetrySecret *v1.Secret)
...
+		}
+		// Check if the endpoint URL needs proxying.
+		proxyConfig := &httpproxy.Config{
+			HTTPProxy:  f.proxy.HTTPProxy(),
+			HTTPSProxy: f.proxy.HTTPSProxy(),
+			NoProxy:    f.proxy.NoProxy(),
+		}
+		proxyFunc := proxyConfig.ProxyFunc()
+		// Ignore errors here.
+		u, _ := url.Parse(rw.URL)
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1444]: An improvement for later (and it can wait for the CMO config CRD) but we could detect invalid URLs and fail early. I suppose that right now, a not parseable URL would fail the prometheus operator reconciliation.

File: pkg/manifests/manifests.go [hunk 4]
```diff
@@ -1424,6 +1425,29 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, telemetrySecret *v1.Secret)
 	for k := range p.Spec.RemoteWrite {
 		rw := &p.Spec.RemoteWrite[k]
 
+		// Fixes: https://issues.redhat.com/browse/OCPBUGS-38289
+		// The code is condensed here to simplify backporting.
+		// This is a temporary workaround because the NoProxy field isn't exported on the Prometheus CRD in older versions.
+		// If the cluster proxy or remoteWrite config change, this code will re-run, ensuring no stale state.
+		// The httpproxy package used here is the same as used by Prometheus, which helps with consistency.
+		if f.proxy.HTTPSProxy() == "" && f.proxy.HTTPProxy() == "" {
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1433]: (nit) we could as well avoid the early continue and always call ProxyFunc()?

File: pkg/manifests/types.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/manifests/types.go:760]: I see several options:
1. keep the current behavior (though it's awkward), if proxyURL is set then use this value, expect if a global HTTP proxy is configured. And it should be explicit in the documentation.
2. Document that the field is deprecated and ignore it.
3. Remove the field (not desired because it could break upgrades).

We could mix 1 and 2: keep the current behavior for a couple of releases and warn users (via the ClusterOperator upgrade condition) that they need to remove it before upgrading to version 4.x (tbd).

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (machine424) [reply pkg/manifests/manifests.go:1454]: Yes, we just need to make sure they use the same definition of a valid URL.
I thought that it would be blocked by the validation webhook, but I'm not sure.

File: pkg/manifests/manifests.go (hunk 2 above)
Comment (machine424) [reply pkg/manifests/manifests.go]: I adjust the code to make it simulate what would happen if CMO set/overwrites the proxy URL here.

File: pkg/manifests/manifests.go (hunk 4 above)
Comment (machine424) [reply pkg/manifests/manifests.go]: I wanted the added logic to be the least intrusive possible.
I adjusted it

File: pkg/manifests/types.go (hunk 1 above)
Comment (machine424) [reply pkg/manifests/types.go:760]: I still think this is a bug, config fields explicitly provided by users should always have precedence, otherwise it makes to sense to keep them exposed.

Let's go with `1.` for now and see.

Comment (simonpasquier) [reply pkg/manifests/types.go:760]: I agree that it's a bug *but* changing the behavior would be a breaking change for existing users.

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/manifests/manifests.go:1454]: which validation webhook?

File: pkg/manifests/manifests.go [hunk 5]
```diff
@@ -1424,12 +1425,41 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, telemetrySecret *v1.Secret)
...
+		// Ignore errors here.
+		u, _ := url.Parse(rw.URL)
+		// Check if the endpoint URL will be using that proxy URL.
+		proxyURL, err := proxyFunc(u)
+		if err == nil && proxyURL == nil {
+			// Proxing will not be used, no need to set it.
+			continue
+		}
+		// The proxy will be used, set it.
+		rw.ProxyURL = tmpRWProxyURL
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1462]:
```suggestion
		rw.ProxyURL = proxyURL.String()
```

File: pkg/manifests/types.go (hunk 1 above)
Comment (machine424) [reply pkg/manifests/types.go:760]: We can always start warning about the current "unusual" behavior and fix that in a future `4.x.0` making sure it's mentioned in the release notes.

(This "unusual" behavior was introduced in 4.16, so it can also be considered as a breaking change for someone upgrading from 4.15 that used to configure the proxyURL via the configmap because they don't want to go through the cluster-wide one.)

But let's continue this discussion somewhere else :)

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (machine424) [reply pkg/manifests/manifests.go:1454]: prom-operator's, but maybe Prometheus CRs do not go through it.

File: pkg/manifests/manifests.go (hunk 5 above)
Comment (machine424) [reply pkg/manifests/manifests.go:1462]: I think it's easier to understand the logic with `tmpRWProxyURL`
(if `noProxy` wouldn't change anything, we fallback to the old behaviour: CMO itself setting `rw.ProxyURL`)

File: Documentation/openshiftdocs/modules/remotewritespec.adoc [hunk 6]
```diff
@@ -35,7 +35,7 @@ link:prometheusrestrictedconfig.adoc[PrometheusRestrictedConfig]
 
 |oauth2|*monv1.OAuth2|Defines OAuth2 authentication settings for the remote write endpoint.
 
-|proxyUrl|string|Defines an optional proxy URL.
+|proxyUrl|string|Defines an optional proxy URL. It is superseded by the cluster-wide proxy if enabled.
```
Comment (machine424) [inline Documentation/openshiftdocs/modules/remotewritespec.adoc:38]: cc @eromanova97, whether the wording is ok for you.

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/manifests/manifests.go:1454]: yeah, the validation webhook only works against PrometheusRule and AlertmanagerConfig objects.

File: pkg/manifests/manifests.go (hunk 5 above)
Comment (simonpasquier) [reply pkg/manifests/manifests.go:1462]: fair enough :)

**Review by simonpasquier — COMMENTED** [review]
nit but commit ec95ac1bf012e0a31a52cc48965ede061455778c has a typo ("proxuying" > "proxying")

**Comment by simonpasquier** [comment]
otherwise lgtm!

File: Documentation/openshiftdocs/modules/remotewritespec.adoc (hunk 6 above)
Comment (eromanova97) [reply Documentation/openshiftdocs/modules/remotewritespec.adoc:38]:
```suggestion
|proxyUrl|string|Defines an optional proxy URL. It is superseded by the cluster-wide proxy, if enabled.
```
Just a very small suggestion :)

**Review by simonpasquier — COMMENTED** [review]
/lgtm
